include/wil/enum_iterator.h

```diff
@@ -0,0 +1,133 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+#ifndef __WIL_ENUM_ITERATOR
+#define __WIL_ENUM_ITERATOR
+
+#include "com.h"
+#include "result_macros.h"
+#include <type_traits>
+#include <tuple>
+
+
+#include <type_traits>
+#include <tuple>
+#include <cstdio>
+
+namespace wil {
+
+    namespace Details {
+        template<typename R, typename T, typename... Args>
+        struct FunctionTraitsBase
+        {
+            using RetType = R;
+            using Type = T;
+            using ArgTypes = std::tuple<Args...>;
+            static constexpr std::size_t ArgCount = sizeof...(Args);
+            template<std::size_t N>
+            using NthArg = std::tuple_element_t<N, ArgTypes>;
+        };
+
+        template<typename F> struct FunctionTraits;
+        
+        template<typename T, typename Arg>
+        struct FunctionTraits<HRESULT(__stdcall T::*)(ULONG, Arg**, ULONG*)>
+            : FunctionTraitsBase<HRESULT, T, ULONG, Arg, ULONG*>
+        {};
+
+#ifdef __cpp_noexcept_function_type
+        template<typename T, typename Arg>
+        struct FunctionTraits<HRESULT(__stdcall T::*)(ULONG, Arg**, ULONG*) noexcept>
+            : FunctionTraitsBase<HRESULT, T, ULONG, Arg, ULONG*>
+        {};
+#endif
+
+        template<typename TEnum>
+        using NextArgType = typename Details::FunctionTraits<decltype(&TEnum::Next)>::template NthArg<1>;
+    }
+
+    template<typename TEnum>
+    /// <summary>
+    /// Iterator class for iterating through types that implement the IEnum* COM pattern.
+    /// </summary>
+    /// <typeparam name="TEnum">The enumerator type, e.g. IEnumIDList, IDiaEnumSymbols.</typeparam>
+    struct enum_iterator{
+        enum_iterator(const enum_iterator& other) = default;
+        enum_iterator(TEnum* pEnum) : m_pEnum(pEnum) {
+            (*this)++;
+        }
+        /// <summary>
+        /// The type of the out parameter in Next(...) that corresponds to the element type being enumerated.
+        /// e.g. PITEMID_CHILD, IDiaSymbol, etc.
+        /// </summary>
+        using TElem = Details::NextArgType<TEnum>;
+        ~enum_iterator() = default;
+
+        auto& operator++(int) { return this->operator++(); }

```

Comment:
I probably have this wrong, but isn't the `int` the # of items to increment? can you pass it to `this->operator++'?

`auto& operator++(int num) { return this->operator++(num); }`

Reply:
It's the postincrement operator (the one without the int is the preincrement operator), i.e. `x++` vs. `++x`.
see [Increment and Decrement Operator Overloading (C++)](https://docs.microsoft.com/cpp/cpp/increment-and-decrement-operator-overloading-cpp)